src/functional/sim_piece.zig

```diff
@@ -0,0 +1,765 @@
+/// `merged_segments_metadata_map`. The `allocator` is used to allocate memory of intermediates.
+fn computeSegmentsMetadataHashMap(
+    merged_segments_metadata: ArrayList(SegmentMetadata),
+    merged_segments_metadata_map: *HashMap(f64, HashMap(
+        f64,
+        ArrayList(usize),
+        HashF64Context,
+        std.hash_map.default_max_load_percentage,
+    ), HashF64Context, std.hash_map.default_max_load_percentage),
+    allocator: mem.Allocator,
+) !void {
+    for (merged_segments_metadata.items) |segment_metadata| {
+        const interception: f64 = segment_metadata.interception;
+        const slope: f64 = (segment_metadata.lower_bound_slope +
+            segment_metadata.upper_bound_slope) / 2;
+
+        // Get or put the outer HashMap entry for the given interception
+        const hash_to_hash_result = try merged_segments_metadata_map.getOrPut(interception);
+        if (!hash_to_hash_result.found_existing) {
+            hash_to_hash_result.value_ptr.* = HashMap(
+                f64,
+                ArrayList(usize),
+                HashF64Context,
+                std.hash_map.default_max_load_percentage,
+            ).init(allocator);
+        }
+        // Get or put the inner HashMap entry for the given slope
+        const hash_to_array_result = try hash_to_hash_result.value_ptr.*.getOrPut(slope);
+        if (!hash_to_array_result.found_existing) {
+            hash_to_array_result.value_ptr.* = ArrayList(usize).init(allocator);
+        }
+        try hash_to_array_result.value_ptr.*.append(segment_metadata.start_time);
+    }
+}
+
+/// Sim-Piece Phase 4. Create from the `merged_segments_metadata_map` and byte array that can be
+/// decoded during decompression and store it in `compressed_values`.
+fn createCompressedRepresentation(
+    merged_segments_metadata_map: HashMap(f64, HashMap(
+        f64,
+        ArrayList(usize),
+        HashF64Context,
+        std.hash_map.default_max_load_percentage,
+    ), HashF64Context, std.hash_map.default_max_load_percentage),
+    compressed_values: *ArrayList(u8),
+) !void {
+    // Iterate over the hash to hash map.
+    var hash_to_hash_iterator = merged_segments_metadata_map.iterator();
+    while (hash_to_hash_iterator.next()) |hash_to_hash_entry| {
+        const current_interception: f64 = hash_to_hash_entry.key_ptr.*;
+        // Append the `current_interception`.
+        try appendValue(f64, current_interception, compressed_values);
+
+        // Append the number of slopes hashed by the `current_interception`.
+        try appendValue(usize, hash_to_hash_entry.value_ptr.*.count(), compressed_values);
+        var hash_to_array_iterator = hash_to_hash_entry.value_ptr.*.iterator();
+
+        while (hash_to_array_iterator.next()) |hash_to_array_entry| {
+            const current_slope = hash_to_array_entry.key_ptr.*;
+            // Append the `current_slope` that approximates the segment.
```

Comment:
Maybe delete as it is clear from the code.

Reply:
Done.